### PR TITLE
Fix cacher test flakiness

### LINKED ISSUE #95394: [Flaky Test] go_test: //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
**Which jobs are flaking**:
bazel-test-master (periodic-kubernetes-bazel-test-master #1314084281143791616)

**Which test(s) are flaking**:
go_test: //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2 

**Testgrid link**:
https://testgrid.k8s.io/google-unit#bazel-test-master
**Reason for failure**:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1314084281143791616

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //staging/src/k8s.io/apiserver/pkg/storage/cacher:go_default_test
-----------------------------------------------------------------------------
W1008 06:36:28.867515  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:34.958950  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:36.960954  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:39.179064  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
W1008 06:36:42.651246  209647 cacher.go:148] Terminating all watchers from cacher *example.Pod
--- FAIL: TestDispatchEventWillNotBeBlockedByTimedOutWatcher (0.11s)
    cacher_whitebox_test.go:1072: watcher is blocked by slower one (count: 22)
W1008 06:36:43.498316  209647 reflector.go:436] k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:503: watch of *v1.Pod ended with: very short watch: k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:503: Unexpected watch close - watch lasted less than a second and no items received
FAIL
```

**Anything else we need to know**:
- links to go.k8s.io/triage appreciated

https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=apiserver%2Fpkg%2Fstorage%2Fcacher%2Fgo_default_test

```


Time | Job | Test
-- | -- | --
08/10/2020, 11:36:02 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
07/10/2020, 20:49:41 | periodic-kubernetes-bazel-test-canary | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
07/10/2020, 03:49:04 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
02/10/2020, 13:06:56 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2
02/10/2020, 10:40:48 | periodic-kubernetes-bazel-test-master | //staging/src/k8s.io/apiserver/pkg/storage/cacher/go_default_test:run_1_of_2


```

- links to specific failures in spyglass appreciated

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1314084281143791616

cc @kubernetes/ci-signal

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix https://github.com/kubernetes/kubernetes/issues/95394

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 1]
```diff
@@ -998,6 +998,15 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
 	wg.Wait()
 }
 
+type fakeTimeBudget struct {}
+
+func (f *fakeTimeBudget) takeAvailable() time.Duration {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1003]: I'm not very familiar with how this is used... what does this duration represent/do?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 2]
```diff
@@ -998,6 +998,15 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
...
+	return 500*time.Millisecond
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1004]: does something actually have to complete within this 500ms or is this just inserting a delay?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go [hunk 3]
```diff
@@ -57,7 +62,7 @@ func newTimeBudget(stopCh <-chan struct{}) *timeBudget {
 	return result
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go:62]: this is orthogonal to this issue, but I was surprised to see an active refresher process here... I expected something more like https://github.com/kubernetes/kubernetes/blob/master/vendor/golang.org/x/time/rate/rate.go#L306-L331 where the budget has an internal clock and computes budget on demand rather than waiting on an active background refresher process at 1-second intervals

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1003]: It's described in the interface definition in time_budget.go.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1004]: it's just giving slack - this whole test should be taking less than a second so it's generally about ensuring that we the channel will never be full when we are trying to insert to it (we wait up to these 500ms for inserting)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go:62]: Filed https://github.com/kubernetes/kubernetes/issues/95875

**Review by wojtek-t — COMMENTED** [review]
@liggitt - PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go]: ok... 500ms still seems like something surprisingly high CPU contention could cause issues with. is the issue that the blocked watcher will use up whatever timeout we give it, so we don't want to give it a really long one like wait.ForeverTestTimeout because it would make the test take forever?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go]: Yes - this test will consume that period fully exactly once, so that's why I don't want to set very high.
I bumped to 2 seconds - let me know if that's ok for you.
